### PR TITLE
Automatic audio resampling

### PR DESCRIPTION
**Pull request by Sami32**
Solved this issue https://github.com/UniversalMediaServer/UniversalMediaServer/issues/588 for me.

---
This change is [Reviewable](https://reviewable.io/reviews/universalmediaserver/universalmediaserver/1000)

### CODE REVIEW

**Comment by Nadahar** [comment]
@Sami32 I still think we're talking past each other. #588 has always been about when NOT using ```AudioResample``` yet most of your changes apply to this configuration.

File: src/main/java/net/pms/encoders/FFmpegAudio.java [hunk 1]
```diff
@@ -211,15 +211,25 @@ public ProcessWrapper launchTranscode(
...
 		}
 
 		if (configuration.isAudioResample()) {
 			if (params.mediaRenderer.isTranscodeAudioTo441()) {
 				cmdList.add("-ar");
 				cmdList.add("44100");
+				cmdList.add("-ac");
+				cmdList.add("2");
+				cmdList.add("-ab");
+				cmdList.add("1411200");
```
Comment (atamariya) [inline src/main/java/net/pms/encoders/FFmpegAudio.java]: You really don't want to hardcode such non-standard values for bitrate - leave alone standard values.

**Comment by Sami32** [comment]
@atamariya Feel free to change it if you like, it's fine for me ;-)

**Comment by Sami32** [comment]
You are all more skilled than me here, so now that you get the idea you can solve it.

Here is one of the part in the code that need to be changed to use `Supported = f:lpcm` in the renderer configuration, in _RendererConfiguration.java_. https://github.com/UniversalMediaServer/UniversalMediaServer/commit/9eb4c59e08fd2193d63765fa89e421dfe0668cd8

**} else if (media != null) {
							AudioProperties audio = media.getFirstAudioTrack().getAudioProperties();
							if (audio.getSampleFrequency() > 0) {
								matchedMimeType += ";rate=" + Integer.toString(audio.getSampleFrequency());
							}
							if (audio.getNumberOfChannels() > 0) {
								matchedMimeType += ";channels=" + Integer.toString(audio.getNumberOfChannels());
							}
						}**

To paraphrase @Nadahar , i will NOT go any further when more competent people are around.

**Comment by Nadahar** [comment]
@Sami32 I think you've misinterpreted what I've said. My reluctance to do things isn't based on inferior skill but by the history in this team.

Quick recap: Earler I tried to fix every error I found, but it turned out this wasn't very popular with the other team members. While I see all code as "common property" that should be improved as much as possible, others feel that the original author "owns" the code they've made and that others shouldn't change that. 

I was encouraged to make my own additions instead of trying to fix other peoples contributions. This made my goals unreachable as I see a lot of problems that need to be fixed before I think it's worth adding new functionality. We also disagree if the current state is "broken" or not. I consider many parts of UMS to be broken and that the only way to handle it is a major restructure or the code, because as it is it's almost impossible to do anything without breaking other things. The others don't agree with.

As the rest of the team don't want me to fix existing bugs and don't want to restructure the things that I consider a mess, there's not a lot I can do, as I find it a waste of time to develop new functionality on top of something "broken".

When it comes to testing/debugging I've always felt that whoever is doing the fixing should do as much of the testing as possible. It's just simpler, you don't get long delays while waiting for communication, you get much less misunderstandings and you can more easily do broader tests. I think there's a bit of a "trend" here that whoever reports a bug has the responsibility to do all the dirty work (testing). I think that's just stupid for the reasons above, plus the fact that it creates an atmosphere where people tend to not report things because then they're stuck with time consuming, boring testing.

About the code snippet I don't understand what you think needs changing. As you can see from the diff the hardcoding of 2 channels for ```isAudioResample``` wasn't done by me, it has probably been like that since the PMS era. As long as the transcoding match this (that is, it will transcode mono or multichannel into stereo) I don't see how that's a problem.

**Comment by Sami32** [comment]
@Thanks for clarifying the things.

I'm too new to have an overview of that and know only @SubJunk , @valib and you, verbose guy :wink: 
I share your opinion on open source code, nothing is our property and do not need a "rooster do battle", male specificity ;-)
I think @SubJunk is afraid to broken things more than they are, i can understand though, but i don't see how clean the code without this risk, and he want clean it as well if i understood correctly.
"to be caught between two stools" like tell the proverb ?

Having a beta tester branch will necessite more server resource, and trying making "sure" the code didn't broke more is time consuming, meaning less updates or less news features; but that is in the hand of the owner of this nice project.
Trully, i don't know what could be the best way to do that, but for sure the code need to be cleaned.
I could be wrong, but he is a little like a diplomat, he need to "spare goat and cabbage": plugin users and plugin makers and gamer users of course !
Maybe making a special cleaned version for beta testers, 1 or 2 times by year, and wait for feedback ?

I apology to overreact, it's my specifity 8-/
So, nobody will cry if we change the code snippets concerning the channels, whatever i was told that cry is good for health ;-)

I hope that the work of @atamariya will come back, and see many merge.
I promise i will test again your work @atamariya , but, pleeaaaase, with little chunks :wink: we are not all skilled as you...as you saw already your work was too scary ;-)

Personnally, when i report a bug i feel like i need search on that, later if necessary, to help at a faster or easier solving, that said there is no obligation to do it, but as developper or skilled people it will be nice.

This is how i see the work need to be done, i could be wrong though.
 That's said my Java skill is not suffisant to continue this PR, it's why i told anyone can modify it, this one or any other one, i'll really not mind.
`audio_resample=true` : Make the mono channel if detected by MediaInfo otherwide 2 channels. "-ac" will need to be set accordingly to this.
`audio_resample=false` : Make the channel and rate take the informations from the renderer configuration if there is or use MediaInfo informations. Set the "-ac" and "sampling frequency" parameter accordingly ("sampling frequency" need to be present in renderer conf otherwiser ignored, since some renderer could only accept 88.2 kHz or any other specific rate)
About the "bitrate", that could be also added only if present in the renderer conf, it depend if you and other people are able to play 96 kHz or 192 and 6 or 8 channels. I don't have a such renderer to test this upper bitrate limitation.

IMHO the problem is that some encoding information are not enforced and hardcoded one is not necessary the real one, and when they do not accord together the renderer is fucked up.
As it headless LPCM it need to be send with additionnal informations that need to be correct, real, as it's with this that the renderer play it.

This time i hope that i'm more clear ?

**Comment by Nadahar** [comment]
@Sami32 As stated before, I consider it "scary" to touch a function that I do not know why is there/what purpose it is supposed to serve. Because of that, I've stayed far away from touching ```audio_resample=true```. The whole function might just be a dead end that noone needs or it might fill some specific use case very well. Because I haven't seen any documentation/explanation of why it's done this strange way, it's impossible to know what potential consequences there would be by changing the channels in the mime type. It could be that the specific case it's meant to serve requires 2 channels. I'm guessing that FFmpeg have no problem "upmixing" mono to stereo - that is to copy the mono channel to both channels. The important thing is that the mime parameter is the same as what the transcoding is producing. I haven't looked at that, but you might have? What happens when ```audio_resample=true``` and a mono audio file is transcoded? Is the output mono or stereo?

When it comes to getting bitrate etc from supported lines it's a good idea in theory, but there are some issues. The format for the supported lines is limited, I don't remember the details but I remember that I found that it wasn't sufficient to properly specify the renderer capabilities. For example I've seen several renderer specs where only e.g 22, 44 and 48 kHz is supported. If you specify 48000 in the renderer configuration for such a renderer, anything below 22 kHz will fail. There should be "full range syntax" support, e.g that you can specify:
* from..to
* from..
* ..to
* specific, values
* combinations of the above

However, that kind of changes is bound to break a lot, so it's virtually impossible to get tested and merged. When it comes to anything above 48 kHz I don't know if it's a good idea to send it at all, since there is no DLNA profile that supports it. That means that its unlikely that very many renderers can handle this. Even if the renderer can interpret the file itself if played from USB etc, it doesn't mean that the DLNA software in the renderer expects and handles such non-standard media.

**Comment by Sami32** [comment]
@Nadahar I feel all the time that your making more problems than it is, or my english is not good inhalf.

"-ac 1" ==> channels=1 where is the problem ? as per LPCM and DLNA specification. Except if you tell me that crappy PS3 cannot handle mono ? If it's for listening reason that different...why not, that could leaved with artificial stereo. It's not a real problem, just a choice, concerning the "audio_resample=true" part only.

When "audio_resample=false" the channels need to check in the renderer conf before to be sure that this renderer doesn't have a specific limitation; and the transcoding parameters should be adapted accordingly (ex: "-ac" "5")

Yes, some renderers are able to handle that, like your 192kHz-16Bit-2ch; the DLNA is a minimum imposed requierement, that manufacturers can improve as specified, autorized in the DLNA standard.

Like 90% of user, i use `audio_resample=true` as it's by default when you install UMS. So i can listen you 192kHz song when my renderer have a 48kHz limitation, like the PS3 it seem.
In your specific case you just to use "Automatic audio resampling" and you can play anything under 22kHz or up to 48kHz ;-) (Mainly up limit affect the renderers)

So, i close the unfinished PR, right ? No problem.

**Comment by Nadahar** [comment]
@Sami32 You're not the first to claim that I complicate things, I see it differently. I prefer to try to think of everything in advance instead of having to do a number of post fixes for the things I didn't think about. It also makes it easier to come up with a good design from the start.

I think the language barrier is creating problems for us. I've never said that the PS3 can't handle mono, I haven't even said that the "resample audio" is for the PS3. I've said that I suspect that it is made with the PS3 in mind, but I don't know. The reason that I think that is that I've seen that you have to choose between 44.1 and 48 kHz in the PS3 configuration, a very strange construct indeed. I don't understand why they can't just let it accept either since both are obviously implemented, but that's the way it is. 

The idea of using the PS3 to play music have never been very appealing to me anyway, so I haven't worried so much about the PS3 but left its assumed "special handling" alone.

You don't have to close to PR for me, didn't you say it fixed something for you? It seems reasonable to me that if the mimetype specification are to stay like it is (hardcoded to 2 channels for "audio resample"), the transcoding itself should do the same, so the last two of your changes are probably a good idea (ac = 2). The first change seems redundant to me, as "s16be" should already specify 16 bits, but I haven't checked the specific behaviour for FFmpeg and if it's possible to end up with a different bit depth than 16 while using "s16be".

**Comment by atamariya** [comment]
>  I see a lot of problems that need to be fixed before I think it's worth adding new functionality. We also disagree if the current state is "broken" or not. I consider many parts of UMS to be broken and that the only way to handle it is a major restructure or the code, because as it is it's almost impossible to do anything without breaking other things

@Nadahar I agree with you here. I'm surprised that there are not as many complaints in the forum. I guess that people simply move to other products instead of taking the pains to report them.

PS3server might be UMS's legacy - but it should not be a burden. The whole reason a fork exists is because there are better possibilities. If we are so afraid to change things for better, one might as well stick with PS3server.

**Comment by Sami32** [comment]
@Nadahar The first and now old DLNA devices have a strick minimum compliance, i.e. 44.1 kHz, so like you told: don't forget some users...

I didn't say that you idea is bad, or you are not good will guy :wink: just that i'm more making thing, test and adjust later. By no mean my way is "better", just my way, for now.

I share "fixing first" philosophy, but i guess that adding new features is a "political", competition choice, as the users want new thing working out of the box; i didn't really checked what the other competitor do, only testing if they work better on some issue.
As debbuging will become more difficult, time consuming, and human being what they are, we prefer get the easy way.

**Comment by Nadahar** [comment]
@Sami32 I think the idea of competition is absurd for OSS. To me a solid, working software is preferrable to one filled with features that never rellay work, and OSS is preferrable by principle as long as it isn't very buggy. A lot of OSS is riddeled with bugs and inconsistancies, I and see that as the challenge for OSS to really make it. OSS backed by commercial companies usually do a lot better, but they also often suffer from paid "premium" solutions and "sabotage" of the free solution so that people will grow tired in the long run and buy "premium". I hate that, so for now I prefer those open source projects that's purely voluntary. I see though that most fail in the long run, even though many also make it.

I think there are three major hurdles for open source software:
* If developers are egoistic and only make sure that it works for them in their environment and doesn't care about normal users.
* The inability to get anyone to do the boring parts like looking for "unexciting" bugs or writing documentation.
* The lack of structure (because noone is "in charge" and can order developers to follow a certain structure or procedure) which will in the end make the code so unstructured and patchy that it's virtually unmaintainable.

I believe that if an open source project can manage these hurdles in a good way, there's not much stopping it. As long as there is structure and documentation, it doesn't really matter if existing developers grow tired and move on, it's easy for new ones to pick it up. When everything is an undocumented mess the opposite is true, and projects just die out.

Avoiding these traps is important to me and means I have some strict principles that's important to me. Those aren't always very compatible with the rest of the team, and is very incompatible with anyone looking for a "quick fix".

**Comment by Sami32** [comment]
@Nadahar Nice points with which i can't disagree.

But we cannot ask to one people to come on it, it need to be a collective agreement, since cleaning, fixing and documenting a "foreign" work with its shortcomings is time consuming, too much for one people to assume, again more on free time.

That's said, we saw what's happened in many open source project and to FFmpeg team as well...

Code is living thing, and we cannot anticipate all the issues possible, in no way, since as you tell every user is different and will use it in his own way, so doing our best should be acceptable.

**P.S.** If you do the coding, i will document it in the wiki, deal ? :wink:

**Comment by Nadahar** [comment]
@Sami32 I'm not sure I understand everything you say, but I agree that ideally this should be a collective agreement. A good first step would be that these things were defined as goals.

It's obviously impossible to anticipate all potential issues, but that doesn't mean one shouldn't try to get the best overview possible. When focusing on one narrow issue/use case while ignoring all others, things are likely to get worse not better. I'm not sure how you define "doing our best", I doubt anyone intend to break anything and "doing their best" can always be applied in that sense. That said, I think that failing to do some research about how a given code impacts other circumstances than the one in focus is very close to negligence. I call this "quality" or "robustness" of code. The quality of the work from the individual coder is just part of the equation though, if there's a lack of structure and logic getting enough overview to make quality code can become a huge task borderlining the impossible. That's why I think cleaning things up is simply a necessity to get somewhere.

**Comment by Sami32** [comment]
@Nadahar So let's do it, and if we encounter problems we could always ask to @Subjunk for a developer's chart that will make things clear and structured for every new and old developper.
Just hoping that we could make it with common sense and good will.
But just don't forget that we use our free time and we need to enjoy it as well :wink:

**Comment by Sami32** [comment]
@Nadahar If you don't worry to post you Onkyo UPnP analyse: (easy and fast ;-))
http://coherence.beebits.net/wiki/UPnP-Inspector#DownloadandInstallation
http://coherence-project.org/download/UPnP-Inspector-0.2.2-1-setup.exe

We should see if your renderer have "LPCM_low" profile implemented.

**Comment by Nadahar** [comment]
I don't know how ```ProtocolInfo``` work but I know that it supports much more formats then what is presented here. There is a source field there which I didn't fill in, maybe I have to give some meaningful input to get some meaningful output. 

From my tests the Onkyo supports all ```LPCM_low``` profiles, but it's not in ```ProtocolInfo```. What is interesting is what is the highest sample rate and/or channels it can accept. It plays ```FLAC``` just fine for example, yet nothing is mentioned about that.

```http-get:*:audio/L16;rate=48000;channels=2:DLNA.ORG_PN=LPCM,http-get:*:audio/L16;rate=48000;channels=1:DLNA.ORG_PN=LPCM,http-get:*:audio/L16;rate=44100;channels=2:DLNA.ORG_PN=LPCM,http-get:*:audio/L16;rate=44100;channels=1:DLNA.ORG_PN=LPCM,http-get:*:audio/wav:DLNA.ORG_PN=WAV,http-get:*:audio/mpeg:DLNA.ORG_PN=MP3,http-get:*:audio/x-ms-wma:DLNA.ORG_PN=WMABASE,http-get:*:audio/x-ms-wma:DLNA.ORG_PN=WMAFULL,http-get:*:audio/x-ms-wma:DLNA.ORG_PN=WMALSL,http-get:*:audio/x-ms-wma:DLNA.ORG_PN=WMDRM_WMABASE,http-get:*:audio/x-ms-wma:DLNA.ORG_PN=WMDRM_WMALSL,http-get:*:audio/vnd.dlna.adts:DLNA.ORG_PN=AAC_ADTS_320,http-get:*:audio/mp4:DLNA.ORG_PN=AAC_ISO_320,http-get:*:audio/3gpp:DLNA.ORG_PN=AAC_ISO_320,```

**Comment by Sami32** [comment]
You should try Magpie as every tools are not identical, as i discovered myself by comparing them; perhaps FLAC Mimetype will appear as well ?

That's said my Panasonic Viera can read Vorbis (2 channels only) in MKV but it's mentionned no where...test and try method give sometimes good surprise ;-)

Morale of the story, automatic detection by ProtocolInfo is good but not infallible...

**P.S.** This time it's your terminology that is not correct ;-)
Your renderer don't have `LPCM_low` profile implemented but support all the specifications.
And that's smart, because as you saw it's was a short lived profile; Sony did implemented if i remember well.

**Comment by Nadahar** [comment]
@Sami32 The ```GetProtocolInfo``` ```Sink``` argument is just a comma separated string. I really doubt that using a different tool will fetch a different string. It's a simple request with no arguments, so I really can't see that the returned result will vary.

I don't remember all the formats I have gotten it to accept, but I know that the list is far greater than what is listed by ```GetProtocolInfo```. The question is if this is unique for the Onkyo, or if renderers generally have a sloppy implementation of this. If so, #987 might be close to pointless.

I'm sorry about the wrong use of terminology, when I said "profiles" I meant the different combinations contained in the ```LPCM_low``` profile.

As stated above, I see no reason why the results returned by #987 or any other software would be different, so I consider that a waste of time. 

Btw, the GTK package bundled with ```UPnP Inspector``` didn't manage to set up my path correctly, so it didn't work. I figured it out in the end (that it was the ```Path``` environment variable that was configured wrong), but you should mention the possibility when recommending that people install it. The installer wasn't very helpful, I got no indication that anything went wrong. When I tried to start ```UPnP Inspector``` is complained about a missing DLL file. The bug itself is quite interesting, I don't know how they managed to make it so, but it was supposed to add ```C:\Program Files (x86)\Common Files\GTK\2.0\bin``` to the path. What it actully added was ```C:\Program Files (x86)\Common Files``` which makes no sense whatsoever since that folder doesn't contain a single file (just subfolders) on my computer.

**Comment by Sami32** [comment]
@Nadahar No need, i already know because i tested myself; i'm a practical guy as the theory can have some limitation :wink: 

About UPnP inspector it was perfectly working on my Windows XP :-D so i didn't imagine that it could give problems. But like tell the proverb, no problems only solutions ;-)
At first i gived you this link http://coherence-project.org/wiki/WikiStart#Download but i later delete it.

**Comment by Nadahar** [comment]
> About your question, you will need to try @atamariya PR yourself to get your answer ;-)

@Sami32 I don't recall asking a question..

**Comment by Sami32** [comment]
@Nadahar 
> The question is if this is unique for the Onkyo, or if renderers generally have a sloppy implementation of this. If so, #987 might be close to pointless.

**Comment by Nadahar** [comment]
@Sami32 That wasn't me asking a question, that was me saying that if other renderers have equally sloppy implementation of ProtocolInfo using it to decide will lead to a lot of unnecessary transcoding and loss of quality. I can't see how I could test this, I don't have those renderers to test.

**Pull request closed without merging**